File: src/tray.cpp
#include <string.h>
#include <iostream>

#include "libnativeapi/include/nativeapi.h"
#include "tray.h"

using namespace nativeapi;

extern TrayManager g_tray_manager;

Tray* __tray(long tray_id) {
  std::shared_ptr<Tray> tray_ptr = g_tray_manager.Get(tray_id);
  if (tray_ptr != nullptr) {
    return tray_ptr.get();
  }
  return nullptr;
}

FFI_PLUGIN_EXPORT
void tray_set_icon(long tray_id, const char* icon) {
  __tray(tray_id)->SetIcon(icon);
}

FFI_PLUGIN_EXPORT
void tray_set_title(long tray_id, const char* title) {
  __tray(tray_id)->SetTitle(title);
}

FFI_PLUGIN_EXPORT
const char* tray_get_title(long tray_id) {
  static std::string title_storage = __tray(tray_id)->GetTitle();
  return title_storage.c_str();
}

FFI_PLUGIN_EXPORT
void tray_set_tooltip(long tray_id, const char* tooltip) {
  __tray(tray_id)->SetTooltip(tooltip);
}

FFI_PLUGIN_EXPORT
const char* tray_get_tooltip(long tray_id) {
  static std::string tooltip_storage = __tray(tray_id)->GetTooltip();
  return tooltip_storage.c_str();
}

FFI_PLUGIN_EXPORT
void tray_start_listening(long tray_id) {
  // TODO: Implement
}

FFI_PLUGIN_EXPORT
void tray_stop_listening(long tray_id) {
  // TODO: Implement
}

File: lib/src/tray.dart
import 'dart:ffi';
import 'dart:ui';

import 'package:ffi/ffi.dart';
import 'package:nativeapi/src/ffi/bindings.dart';
import 'package:nativeapi/src/ffi/bindings_generated.dart';

/// A class for representing a tray.
///
/// This class provides methods for setting the icon, title, and tooltip of a tray.
class Tray {
  /// The native API bindings.
  NativeApiBindings get _bindings => nativeApiBindings;

  /// The ID of the tray.
  final int id;

  /// Creates a new tray.
  Tray({required this.id});

  /// Sets the icon of the tray.
  void setIcon(String icon) {
    _bindings.tray_set_icon(id, icon.toNativeUtf8().cast<Char>());
  }

  /// Sets the title of the tray.
  void setTitle(String title) {
    _bindings.tray_set_title(id, title.toNativeUtf8().cast<Char>());
  }

  /// Gets the title of the tray.
  String getTitle() {
    return _bindings.tray_get_title(id).cast<Utf8>().toDartString();
  }

  /// Sets the tooltip of the tray.
  void setTooltip(String tooltip) {
    _bindings.tray_set_tooltip(id, tooltip.toNativeUtf8().cast<Char>());
  }

  /// Gets the tooltip of the tray.
  String getTooltip() {
    return _bindings.tray_get_tooltip(id).cast<Utf8>().toDartString();
  }

  void addListener(VoidCallback listener) {}

  @override
  int get hashCode => id.hashCode;

  @override
  operator ==(Object other) {
    return other is Tray && other.id == id;
  }
}
